nexus/src/app/sagas/mod.rs

```diff
@@ -116,3 +116,129 @@ pub(super) async fn saga_generate_uuid<UserType: SagaType>(
 ) -> Result<Uuid, ActionError> {
     Ok(Uuid::new_v4())
 }
+
+macro_rules! __stringify_ident {
+    ($i:ident) => {
+        stringify!($i)
+    };
+}
+
+macro_rules! __emit_action {
+    ($node:ident, $output:literal) => {
+        paste::paste! {
+            #[allow(dead_code)]
+            fn [<$node:lower _action>]() -> ::steno::Node {
+                ::steno::Node::action(
+                    $output,
+                    crate::app::sagas::__stringify_ident!([<$node:camel>]),
+                    $node.as_ref(),
+                )
+            }
+        }
+    };
+}
+
+macro_rules! __action_name {
+    ($saga:ident, $node:ident) => {
+        paste::paste! {
+            concat!(
+                stringify!($saga),
+                ".",
+                crate::app::sagas::__stringify_ident!([<$node:lower>]),
+            )
+        }
+    };
+}
+
+/// A macro intended to reduce boilerplate when writing saga actions.
+///
+/// For this input:
+///
+/// ```ignore
+/// declare_saga_actions! {
+///     my_saga;
+///     SAGA_NODE1 -> "output1" {
+///         + do1
+///         - undo1
+///     }
+///     SAGA_NODE2 -> "output2" {
+///         + do2
+///     }
+/// }
+/// ```
+///
+/// We generate the following:
+/// - For `SAGA_NODE1`:
+///     - A `NexusAction` labeled "my_saga.saga_node1" (containing "do1" and "undo1").
+///     - `fn saga_node1_action() -> steno::Node` referencing this node, with an
+///     output named "output1".
+/// - For `SAGA_NODE2`:
+///     - A `NexusAction` labeled "my_saga.saga_node2" (containing "do2").
+///     - `fn saga_node2_action() -> steno::Node` referencing this node, with an
+///     output named "output2".
+/// - For `my_saga`:
+///     - `fn my_saga_register_actions(...)`, which can be called to implement
+///     `NexusSaga::register_actions`.
+macro_rules! declare_saga_actions {
+    // The entrypoint to the macro.
+    // We expect the input to be of the form:
+    //
+    //  saga-name;
+    ($saga:ident; $($tail:tt)*) => {
+        declare_saga_actions!(S = $saga <> $($tail)*);
+    };
+    // Subsequent lines of the saga action declaration.
+    // These take the form:
+    //
+    //  ACTION_NAME -> "output" {
+    //      + action
+    //      - undo_action
+    //  }
+    //
+    // However, we also want to propagate the Saga structure and collection of
+    // all node names, so this is *actually* parsed with a hidden prefix:
+    //
+    //  S = SagaName <old nodes> <> ...
+    //
+    // Basically, everything to the left of "<>" is just us propagating state
+    // through the macro, and everything to the right of it is user input.
+    (S = $saga:ident $($nodes:ident),* <> $node:ident -> $out:literal { + $a:ident - $u:ident } $($tail:tt)*) => {
+        lazy_static::lazy_static! {
+            static ref $node: crate::app::sagas::NexusAction = ::steno::ActionFunc::new_action(
+                crate::app::sagas::__action_name!($saga, $node), $a, $u,
+            );
+        }
```

Comment:
Can we use `once_cell::Lazy` here (maybe in a followup since this just refactors code?) `lazy_static` is mostly deprecated at this point I think.

Reply:
Done in 6abce0865dfb225b3581f05df7a544b2561168f8 ! We are using `lazy_static` elsewhere in Omicron, but that can be converted too